### PR TITLE
Added support to srv generation from multiple sources/namespaces to opensplice

### LINKED ISSUE #6: Allow generating messages services in an `action` folder
#### Feature description

`rosidl_typesupport_opensplice_c` and `rosidl_typesupport_opensplice_cpp` need to allow generating messages and services in an `action` folder and namespace

#### Implementation considerations

The service templates would need to change to allow service definitions to live in the .actions folder

https://github.com/ros2/rosidl_typesupport_opensplice/blob/70ee746f077d56b7c7e3e1aa51a35f80b10bd1e9/rosidl_typesupport_opensplice_c/resource/srv__type_support_c.cpp.em#L25-L30

https://github.com/ros2/rosidl_typesupport_opensplice/blob/70ee746f077d56b7c7e3e1aa51a35f80b10bd1e9/rosidl_typesupport_opensplice_cpp/resource/srv__type_support.cpp.em#L29-L34

connects to ros2/rosidl#301

### PR DESCRIPTION
**Pull request by sservulo**
Closes #6 . Connected to ros2/rosidl#301. This pull request is part of a series of pull requests that make the necessary changes to allow us to build .srv files outside an srv folder.

connects to ros2/rosidl#301

### CODE REVIEW

File: rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake [hunk 1]
```diff
@@ -70,20 +70,27 @@ foreach(_idl_file ${rosidl_generate_interfaces_IDL_FILES})
...
       "${_output_path}/${_parent_folder}/dds_opensplice_c/${_header_name}__type_support_c.cpp")
   elseif(_extension STREQUAL ".srv")
-    list(APPEND _generated_srv_files "${_output_path}/srv/${_header_name}__rosidl_typesupport_opensplice_c.h")
-    list(APPEND _generated_srv_files "${_output_path}/srv/dds_opensplice_c/${_header_name}__type_support_c.cpp")
+    get_filename_component(_parent_folder "${_idl_file}" DIRECTORY)
+    get_filename_component(_parent_folder "${_parent_folder}" NAME)
+    if(_parent_folder STREQUAL "srv")
+      set(_var "_generated_srv_files")
+    else()
+      message(FATAL_ERROR "Interface file with unknown parent folder: ${_idl_file}")
```
Comment (sloretz) [inline rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake:67]: I get an error here when generating a service in an `action` folder

```
CMake Error at /home/developer/workspaces/ros2-dev0/install/rosidl_typesupport_opensplice_c/share/rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake:57 (message):
  Interface file with unknown parent folder:
  /home/developer/workspaces/ros2-dev0/src/ros2/rcl_interfaces/test_msgs/action/ShaneDynamicArrayPrimitivesNested.msg
Call Stack (most recent call first):
  /home/developer/workspaces/ros2-dev0/install/ament_cmake_core/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:38 (include)
  /home/developer/workspaces/ros2-dev0/install/rosidl_cmake/share/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake:186 (ament_execute_extensions)
  CMakeLists.txt:42 (rosidl_generate_interfaces)
```

Comment (sloretz) [reply rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake:67]: Oops, the error is from line 57 above. I don't think a middleware specific generator needs to check if the parent folder is known or not, so both here and above that restriction could be removed.

File: rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake [hunk 2]
```diff
@@ -77,20 +77,27 @@ foreach(_idl_file ${rosidl_generate_interfaces_IDL_FILES})
...
       "${_output_path}/${_parent_folder}/dds_opensplice/${_header_name}__type_support.cpp")
   elseif(_extension STREQUAL ".srv")
-    list(APPEND _generated_srv_files "${_output_path}/srv/${_header_name}__rosidl_typesupport_opensplice_cpp.hpp")
-    list(APPEND _generated_srv_files "${_output_path}/srv/dds_opensplice/${_header_name}__type_support.cpp")
+    get_filename_component(_parent_folder "${_idl_file}" DIRECTORY)
+    get_filename_component(_parent_folder "${_parent_folder}" NAME)
+    if(_parent_folder STREQUAL "srv")
+      set(_var "_generated_srv_files")
+    else()
+      message(FATAL_ERROR "Interface file with unknown parent folder: ${_idl_file}")
```
Comment (sloretz) [inline rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake:74]: Same as above, this errors when a service is in an `action` folder.

**Comment by sloretz** [comment]
A few more places with `srv` hard coded

https://github.com/ros2/rosidl_typesupport_opensplice/blob/8a9650cd589c967c56d5ce2f70b617646661acf6/rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake#L118

https://github.com/ros2/rosidl_typesupport_opensplice/blob/8a9650cd589c967c56d5ce2f70b617646661acf6/rosidl_typesupport_opensplice_c/cmake/rosidl_typesupport_opensplice_c_generate_interfaces.cmake#L123

https://github.com/ros2/rosidl_typesupport_opensplice/blob/8a9650cd589c967c56d5ce2f70b617646661acf6/rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake#L142

https://github.com/ros2/rosidl_typesupport_opensplice/blob/8a9650cd589c967c56d5ce2f70b617646661acf6/rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake#L147

**Comment by sloretz** [comment]
A couple more `srv` in 

https://github.com/ros2/rosidl_typesupport_opensplice/pull/7/files#diff-e9f8233c24a2a2da70d5d4691bbf1089L107

and 

https://github.com/ros2/rosidl_typesupport_opensplice/pull/7/files#diff-e9f8233c24a2a2da70d5d4691bbf1089L116

**Review by sloretz — APPROVED** [review]
Works for me with python and c++ clients/services/publishers/subscribers using opensplice as the middleware impelementation

**Comment by apojomovsky** [comment]
Ok, given that we have green light I'm proceeding to merge this and its related PRs

**Pull request merged**